File: src/menu/menu_lose.c
#include "menu.h"

#include "../game_loop/game_loop.h"
#include "../gui/gui.h"
#include "../keyboard_manager/keyboard_manager.h"
#include "../util/util.h"

static const char *buttons[] = {
  "RESTART GAME",
  "EXIT",
};

constexpr i32 length = ArrayLength(buttons);

void MenuLoseTick() {
  HandleUpDownSelection(&loseMenu, length);

  if (SingleKeyPress(KeyEnter)) {
    switch (loseMenu.selectedIndex) {
    case 0: // Restart game
      Restart();
      break;

    case 1: // Exit
      menuID = MenuExit;
      prevMenuID = MenuLose;
      break;
    }
  }
}

void MenuLoseRender() {
  RenderString(0, 0, 40, 0xFF, 0xFF, 0x00, 0xFF, 1, -3, "YOU LOSE!");
  RenderString(0, 0, 40, 0xFF, 0xFF, 0x00, 0xFF, 1, -2, "TRY AGAIN?");

  for (i32 i = 0; i < length; i++) {
    if (loseMenu.selectedIndex == i) {
      RenderString(
        0, 0, 40, 0xA0, 0xA0, 0x00, 0xFF, 1, i, "> %s <", buttons[i]);
    } else {
      RenderString(0, 0, 40, 0xFF, 0xFF, 0x00, 0xFF, 1, i, "%s", buttons[i]);
    }
  }
}

File: menu_lose.go
package main

// loseMenu is the game over menu navigation state.
var loseMenu Menu

// loseButtons are the menu options after losing.
var loseButtons = []string{"RESTART GAME", "EXIT"}

// menuLoseTick handles input for the game over screen.
func menuLoseTick() {
	handleUpDownSelection(&loseMenu, len(loseButtons))

	if SingleKeyPress(KeyEnter) {
		switch loseMenu.SelectedIndex {
		case 0: // Restart game
			restart()
		case 1: // Exit
			menuID = MenuTypeExit
			prevMenuID = MenuTypeLose
		}
	}
}

// menuLoseRender draws the game over screen.
func menuLoseRender() {
	RenderStringf(
		RenderStringOptions{
			Size:                   menuFontSize,
			Color:                  menuNormalTextColor,
			RelativeToWindowCenter: true,
			LineNo:                 -3,
		},
		"YOU LOSE!",
	)

	RenderStringf(
		RenderStringOptions{
			Size:                   menuFontSize,
			Color:                  menuNormalTextColor,
			RelativeToWindowCenter: true,
			LineNo:                 -2,
		},
		"TRY AGAIN?",
	)

	for i, button := range loseButtons {
		opts := RenderStringOptions{
			Size:                   menuFontSize,
			Color:                  menuNormalTextColor,
			RelativeToWindowCenter: true,
			LineNo:                 i,
		}

		format := "%s"

		if loseMenu.SelectedIndex == i {
			opts.Color = menuHoverTextColor
			format = "> %s <"
		}

		RenderStringf(opts, format, button)
	}
}
